From 81559ce1313941d55110571b8683f1202bdbe5fa Mon Sep 17 00:00:00 2001
From: Jeff Sharkey <jsharkey@android.com>
Date: Thu, 16 Apr 2020 12:22:45 -0600
Subject: [PATCH 15/26] Handle extractVolumePath() being null.

In some situations, the data path being migrated may not actually
be a valid /storage/ path, in which case we should gracefully
continue the migration flow without attempting a rename for pending
or trashed items.

Bug: 153825968
Test: atest --test-mapping packages/providers/MediaProvider
Change-Id: I3e4808b369f34d11797793e938ad61596dbc60fe
(cherry picked from commit 635389aed6d645e8c07eaf656ff5ff3b415c9e4c)
---
 .../providers/media/DatabaseHelper.java       | 24 ++++++++++---------
 1 file changed, 13 insertions(+), 11 deletions(-)

diff --git a/src/com/android/providers/media/DatabaseHelper.java b/src/com/android/providers/media/DatabaseHelper.java
index afa3d637..256c1870 100644
--- a/src/com/android/providers/media/DatabaseHelper.java
+++ b/src/com/android/providers/media/DatabaseHelper.java
@@ -809,17 +809,19 @@ public class DatabaseHelper extends SQLiteOpenHelper implements AutoCloseable {
 
                     // When migrating pending or trashed files, we might need to
                     // rename them on disk to match new schema
-                    FileUtils.computeDataFromValues(values,
-                            new File(FileUtils.extractVolumePath(data)));
-                    final String recomputedData = values.getAsString(MediaColumns.DATA);
-                    if (!Objects.equals(data, recomputedData)) {
-                        try {
-                            Os.rename(data, recomputedData);
-                        } catch (ErrnoException e) {
-                            // We only have one shot to migrate data, so log and
-                            // keep marching forward
-                            Log.w(TAG, "Failed to rename " + values + "; continuing");
-                            FileUtils.computeValuesFromData(values);
+                    final String volumePath = FileUtils.extractVolumePath(data);
+                    if (volumePath != null) {
+                        FileUtils.computeDataFromValues(values, new File(volumePath));
+                        final String recomputedData = values.getAsString(MediaColumns.DATA);
+                        if (!Objects.equals(data, recomputedData)) {
+                            try {
+                                Os.rename(data, recomputedData);
+                            } catch (ErrnoException e) {
+                                // We only have one shot to migrate data, so log and
+                                // keep marching forward
+                                Log.w(TAG, "Failed to rename " + values + "; continuing");
+                                FileUtils.computeValuesFromData(values);
+                            }
                         }
                     }
 
-- 
2.35.1

